crates/data_components/src/s3_vectors/put_vectors_sink.rs

```diff
@@ -0,0 +1,428 @@
+                    )
+                    .await
+                    .map_err(SdkError::into_service_error)
+                    .context(PutVectorsSnafu)?;
+
+                count += chunk.len();
+            }
+        }
+
+        Ok(count as _)
+    }
+}
+
+#[allow(clippy::result_large_err)]
+fn create_put_input_vectors(record_batch: &RecordBatch) -> Result<Vec<PutInputVector>> {
+    let name = S3_VECTOR_PRIMARY_KEY_NAME.to_string();
+    let keys = record_batch
+        .column_by_name(&name)
+        .ok_or_else(|| Error::MissingColumn { name: name.clone() })?
+        .as_any()
+        .downcast_ref::<arrow::array::StringArray>()
+        .ok_or_else(|| Error::ColumnTypeMismatch {
+            name,
+            expected: "StringArray".to_string(),
+        })?;
+
+    let name = S3_VECTOR_EMBEDDING_NAME.to_string();
+    let vectors = record_batch
+        .column_by_name(&name)
+        .ok_or_else(|| Error::MissingColumn { name: name.clone() })?
+        .as_any()
+        .downcast_ref::<arrow::array::ListArray>()
+        .ok_or_else(|| Error::ColumnTypeMismatch {
+            name,
+            expected: "ListArray".to_string(),
+        })?;
+
+    let schema = record_batch.schema();
+    let fields = schema
+        .fields()
+        .iter()
+        .enumerate()
+        .filter(|(_, f)| {
+            f.name() != S3_VECTOR_EMBEDDING_NAME && f.name() != S3_VECTOR_PRIMARY_KEY_NAME
+        })
+        .map(|(i, f)| (i, f.name(), f.data_type()))
+        .collect::<Vec<_>>();
+
+    let mut put_input_vectors = Vec::new();
+    for row in 0..record_batch.num_rows() {
+        let key = keys.value(row).to_string();
+
+        let vector = vectors
+            .value(row)
+            .as_any()
+            .downcast_ref::<arrow::array::Float32Array>()
+            .ok_or_else(|| Error::ColumnTypeMismatch {
+                name: format!("data[{row}]"),
+                expected: "Float32Array".to_string(),
+            })?
+            .values()
+            .to_vec();
+
+        if vector.iter().any(|&x| x.is_nan() || x.is_infinite()) {
+            continue;
+        }
+
+        if vector.iter().all(|&x| x == 0.0) {
+            continue;
+        }
+
+        let mut metadata = HashMap::new();
+
+        for (index, name, data_type) in &fields {
+            let col = record_batch.column(*index);
+            let value = metadata_from_row(row, data_type, col);
+            metadata.insert((*name).to_string(), value);
+        }
+
+        let metadata = if metadata.is_empty() {
+            None
+        } else {
+            Some(Document::Object(metadata))
+        };
+
+        let put_input_vector = PutInputVector::builder()
+            .key(key)
+            .data(VectorData::Float32(vector))
+            .set_metadata(metadata)
+            .build()
+            .context(BuildInputSnafu)?;
+
+        put_input_vectors.push(put_input_vector);
+    }
+
+    Ok(put_input_vectors)
+}
+
+#[allow(clippy::unwrap_used)]
```

Comment:
We shouldn't use unwrap in production code - if you are confident this can never happen, then we should use the `unreachable!` macro with the reason why it can't be reached.

Reply:
But in this instance, I'd prefer to have this just return a Result instead